### ISSUE TITLE
Possibility to expose more of the Axios original http response?

### ISSUE THREAD

**Issue by rizdaprasetya** [hoppscotch/hoppscotch-extension#41]
Thanks for this useful extensions, you guys rocks! 🤘 

So I've been using this extension outside of official Hoppscotch web app. This extensions work wonderfully for working-around Browser's AJAX CORS limitation, by routing HTTP request through this extension. However I found **some limitation in terms of the exposed HTTP response data**. My use case is to:
- Get the final URL of the response, after an HTTP request encountered HTTP redirect(s).
- I figure [it will need](https://stackoverflow.com/a/62901003) `response.request.responseURL` from the original HTTP response of Axios. Which currently this extension does not expose.

I track it down into these lines, the extension currently return limited set of properties of the Axios's original response object:
https://github.com/hoppscotch/hoppscotch-extension/blob/427969d278ec969cc7091aabc4acaf716cfae643/src/index.ts#L153-L160

Any chance to add maybe something like this?
```javascript
rawAxiosResponse: res

// or maybe:
responseUrl: res.request.responseURL
```

I know that this is your extension mainly tailored for your own web app, so if it is not possible that is completely understandable. I'll probably just make my own fork and build an alternative extension.
Thanks

**Comment by AndrewBastin** [comment]
Hi Rizda!

Thanks for your comments! Happy to see you love the extension!

As you said, the extension only exposes the minimal data because Hoppscotch only needs that few bits of data in the app. But that being said, we might need to eventually add the redirect URL as that is something the app should eventually expose (cc @liyasthomas) so I am pretty sure we will expose that at some point, but not really soon as we are busy with some stuff in Hoppscotch, so this will just be in the backlog for a while. But if you want to, you can shoot a PR and I will review and merge it into the extension or you can fork off the project and go in your direction.

By the way, I am curious to know what you are doing with the extension outside the Hoppscotch web app, that is if you don't mind sharing. If you want to share privately, you can ping on my email (it is available on my Github profile page).

Happy hacking ^_^

**Comment by rizdaprasetya** [comment]
Thanks for your kind respond.

Ah yes, the hoppscotch.io (i'll refer as Postwoman, as it is much easier to type TBH 😂 ) could also benefit from the exposed responseURL after redirect. 
<details>
<summary><b>Postwoman use case</b> (click to expand)</summary>
<hr>

I notice that on Postwoman `test script` feature compared to Postman, it's missing a `request` object
On Postman there is `pm.request` object. That is useful on [scenario of](https://community.postman.com/t/when-redirected-how-to-see-the-url-you-re-redirected-to/6699):
- As a Postman user, when HTTP request is getting redirected to other url(s), I should be able to retrieve the final response url
 
So that would be nice to have similar feature on Postwoman:
- As a Postwoman user, when HTTP request is getting redirected to other url(s), I should be able to retrieve the final response url. At least via test script, exposed as  maybe `pw.response.responseURL` (exposed from Axios's original `response.request.responseURL` data).
  - Can be better if it can also be shown in Postwoman UI, but Postman itself also don't show it on UI, so it may not be required.

What that scenario be useful for? For example most of login pages will do 302 redirect, after a successful login request (POST request with username & password on the body).

**Note:** this maybe deserve its own issue/discussion on the main app repo, i'm putting it here so we can copy paste there later.
<hr>
</details>

Anyway back to your comment, yes no rush, this can be put in backlog. I'll also try to open PR.

Sure I can try to share my use case. In summary:
- I'm trying to build a web (let's call it **automator.com**) to automate some task on other web (**somedashboard.com**).
- in order for automator.com to able to perform action (including login, etc.) on somedashboard.com on behalf of the user, automator.com will need to perform HTTP request to somedashboard.com
  - somedashboard.com has no official API, i'll have to do it in a hacky way.
  - user will need to input their password (of somedashboard.com) into automator.com frontend.
  - to keep privacy, user's password (and any of thier data) of somedashboard.com should not be exposed to automator.com backend (or any other party). So it will need to stay in user's browser.
  - any HTTP request attempt to somedashboard.com from automator.com directly via frontend (browser fetch/XHR request) is not possible, due to the somedashboard.com is a cross origin request, thus browser will block the CORS request as per security measure.
    - I could theoretically spin up CORS proxy on automator.com backend, but that will expose sensitive information. So this is not an option.
  - Postwoman extension is built to work-around CORS limitation, so it will be a good solution. Ask customer to install it, whitelist automator.com. Then automator.com will be able to perform frontend browser fetch/XHR request to perform login to somedashboard.com. Customer data stays in their own browser, and automator.com able to perform action to somedashboard.com, get the response and show to user. Win-win.
- But, there is one pesky action that is "login", somedashboard.com does not present an important data that is (let's say) `user_id` on the HTTP response. Login action will result in 302 redirect and the `user_id` only present on the URL after the redirect (e.g: somedashboard.com/timeline/uid123456/home). It shown only once there, and not retrievable anywhere else I try to find.
- So the only way to get the `user_id` is to figure is to retrieve: the final URL after the redirect happen, when Postwoman is used to proxy the login HTTP request.
- Unfortunately, Postwoman extension is not yet exposing it anywhere. Hence I open this question here.

Well, this is very specific use case that maybe very few people will encounter. But hopefully that answer your curiosity.

The fact that this Postwoman Extension can help with browser CORS blocking limitation, open up many possibilities for people to use it as proxy to do API/HTTP request they couldn't do before from frontend. It feels like giving frontend only static site a backend capability in terms of API call freedom. Which is cool.

In case anyone wondering how to use Postwoman extension as HTTP request proxy, here's a sample of mine:
```javascript
// perform the http request
let httpResponse = await window.window.__POSTWOMAN_EXTENSION_HOOK__.sendRequest({
    credentials: true,
    wantsBinary: false,
    method: "POST",
    url: "https://somedashboard.com/login",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    data: "username=user&password=pass&execution=someToken&_eventId=submit&geolocation="
});
// log the result
console.log("the httpResponse:",httpResponse);
```

**Comment by rizdaprasetya** [comment]
PR opened, please check, thanks!

**Comment by AndrewBastin** [comment]
Awesome, I will try to review this tomorrow and look into it ^_^

Also thanks for the explanation.

> Postwoman extension is built to work-around CORS limitation, so it will be a good solution. Ask customer to install it, whitelist automator.com. Then automator.com will be able to perform frontend browser fetch/XHR request to perform login to somedashboard.com. 

I suggest you fork off our repo and make your own extension so as to provide the distinction. I don't take the security concerns raised by the extension bypassing CORS lightly and mixing Hoppscotch extension to a totally unrelated website might be bad for both your users and ours. Especially in case we plan to update the extension protocols suddenly in the future and you miss it, it might break your application and affect your users.

**Comment by rizdaprasetya** [comment]
Yes, make sense. I also considered that risk, so I better use a fork instead for my project. This whole issue is just notifying the upstream (this repo) about changes I made and if it may benefit the upstream. The final decision is still understandably upstream project owner's to take, there is no wrong decision there, and no deadline/timeline involved, suit yourself. Cheers 😃

**Comment by AndrewBastin** [comment]
Closed via #42 

I will still update this issue until the feature becomes live in the Chrome and Firefox extension stores.
